### PR TITLE
add ci for asv benchmark suite

### PR DESCRIPTION
**Pull request by jaimergp**
# Description

This PR implements a CI workflow to run `asv` in _continuous_ mode; otherwise known as "relative benchmarking". The idea is that CI environments are too noisy to do absolute benchmarking (throttling, several jobs running at the same time in the same physical hardware), but they should be ok if you compare the performance of one commit against other.

This was implemented for [`scikit-image`](https://github.com/scikit-image/scikit-image/blob/9bdd010a8/.github/workflows/benchmarks.yml#L1) a while ago, as explained in [this blog post](https://labs.quansight.org/blog/2021/08/github-actions-benchmarks/). Since this can potentially take a bit to run, it will only be triggered by the addition of a particular label, not with every commit.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Related to https://github.com/napari/napari/issues/1662

# How has this been tested?

- [x] Benchmarks run successfully
- [x] ... being triggered by a label
- [ ] and artifacts with results can be downloaded for local inspection

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
I think that the Qt part should be benchmarked separately from non-Qt part. There are some problems with waiting on signals that may impact measurements (for example update of the runner version may impact this base on my observations).

**Comment by andy-sweet** [comment]
Super excited to see see this @jaimergp ! What does it take to publish these to a dashboard like https://pandas.pydata.org/speed/pandas/ ?

**Comment by jaimergp** [comment]
It's surprisingly simple yet not trivial! The short story is that it needs to run a dedicated machine. No upgrades, nothing else running. So the difficult part is providing a machine with permanent internet access. Usually, an old machine in somebody's basement :)

**Comment by Czaki** [comment]
> It's surprisingly simple yet not trivial! The short story is that it needs to run a dedicated machine.

@tlambert03 do it for psygnal using github actions https://www.talleylambert.com/psygnal/ so maybe it is not required?

**Comment by tlambert03** [comment]
One thing that makes it much easier to auto-publish the full suite like that for psygnal is that the benchmarks run _much_ faster than they do here.  So I can realistically run benchmarks for all releases in [around 2 hours](https://github.com/tlambert03/psygnal/actions/runs/2328124587) ... meaning i can do it on GHA, and not store anything except the published html.    after starting with @jaimergp's awesome blog post linked above, I modeled [psygnal's CI action](https://github.com/tlambert03/psygnal/blob/6e750e06b9971e997b53db5782ea3a0f78176e53/.github/workflows/benchmark.yml#L13) after one I found in the [tqdm repo](https://github.com/tqdm/tqdm/blob/4f208e72552c4d916aa4fe6a955349ee8b2ed353/.github/workflows/check.yml#L26)

**Comment by tlambert03** [comment]
just want to say i think this is a super important undertaking.  Thank you _so_ much @jaimergp for digging into this

**Comment by jaimergp** [comment]
"Quick" checks with `asv` (just to make sure code in the benchmark suite does not rot and breakages are fixed) takes one hour or so on CI, mostly due to the parameterisation. I am adding some custom skipping mechanism to only run the first param (which should be enough to check API breakage). Locally this takes ~20 mins, but let's see what happens on CI.

**Comment by jaimergp** [comment]
Hm, maybe [this part](https://github.com/napari/napari/blob/5b8e5705882578666e32b1a016839fd85ac87e17/napari/benchmarks/__init__.py#L11) of the "quick skip" logic is not skipping everything (looking at multi-value parameters). I would have expected something faster...

**Comment by jaimergp** [comment]
Hm, it takes also one hour 🤷 

I guess it doesn't make sense to have "quick" checks if they are not quick. API breakage will have to be detected once in `main`.

**Comment by jaimergp** [comment]
I am increasing the timeout (60 by default) to allow slow tests to run.

I am seeing two other issues:

1. [x] `benchmark_text_manager.TextManagerSuite` [tries to use `None`](https://github.com/napari/napari/blob/e5345e76db014434e8418c8aee68509979c56911/napari/benchmarks/benchmark_text_manager.py#L18) as the `string` for `TextManager` but this doesn't seem to be supported anymore?
2. [x] Some out of bounds error because the [passed indices](https://github.com/napari/napari/blob/e5345e76db014434e8418c8aee68509979c56911/napari/benchmarks/benchmark_text_manager.py#L16) seem to be off by one (e.g. `IndexError: index 8 is out of bounds for axis 0 with size 8` in [this one](https://github.com/napari/napari/blob/e5345e76db014434e8418c8aee68509979c56911/napari/benchmarks/benchmark_text_manager.py#L53))

I might need some help here to confirm which strategy is correct, @napari/core-devs:

1. `TextManager` should never receive `None`, so we remove it from the suite. Alternatively, there's another "empty" value we can use, e.g. `""`?
2. Apply `indices - 1` to all the problematic tests?

**Comment by andy-sweet** [comment]
> 1. `TextManager` should never receive `None`, so we remove it from the suite. Alternatively, there's another "empty" value we can use, e.g. `""`?

Sorry for that. And good catch. You can replace `None` with `{'constant': ''}`, but I'd prefer to just remove it.

**Comment by andy-sweet** [comment]
> 2\. Some out of bounds error because the [passed indices](https://github.com/napari/napari/blob/e5345e76db014434e8418c8aee68509979c56911/napari/benchmarks/benchmark_text_manager.py#L16) seem to be off by one (e.g. `IndexError: index 8 is out of bounds for axis 0 with size 8` in [this one](https://github.com/napari/napari/blob/e5345e76db014434e8418c8aee68509979c56911/napari/benchmarks/benchmark_text_manager.py#L53))

I think this is the same underlying cause described here: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/ASV.20benchmark.20fails.20on.20.60run.60.20but.20.20not.20.60dev.60

I'm not sure how to resolve this other than removing or disabling the `time_remove_as_batch` benchmark for now. At least on CI.

**Comment by jaimergp** [comment]
Oh, wow, I was in that thread 🤦  The good news is that I found out the issue: the warmup runs! Last commit fixes it, hopefully. Thank you so much @andy-sweet!

**Comment by jaimergp** [comment]
@jni I think I have figured out a way to have a comment posted to the PR. It requires a 2nd workflow that is triggered upon completion of the `Benchmarks` one. This takes the artifact, prepares the data and posts the comment. The only requirement is that the 2nd workflow runs from `main` (which requires merging this PR) to have the appropriate token.

I think we can get a quick review here, merge the progress so far, and polish the rough edges as we go. Since we barely touch the napari source itself, this should have little impact on the work the rest of the team is doing.

Let me know what you think!

File: .github/workflows/benchmarks.yml
Comment (andy-sweet) [inline .github/workflows/benchmarks.yml:29]: Optional: these explanatory comments are already great, but I might also add a link to the `asv continuous` command for more details: https://asv.readthedocs.io/en/stable/commands.html#asv-continuous

That also helps clarify that these are options for `asv continuous` rather than `asv run` here.

File: napari/benchmarks/benchmark_text_manager.py
Comment (andy-sweet) [inline napari/benchmarks/benchmark_text_manager.py]:
```suggestion
    # occasions: warmup and timing itself. We disable warmup and only
```

File: .github/workflows/test_comprehensive.yml
Comment (andy-sweet) [inline .github/workflows/test_comprehensive.yml]: Any reason why you chose macos-latest compared to one of other runners? If so, might want to add an explanatory comment. Especially as ubuntu-latest is used for the benchmark itself. I feel like our default runner for most things is ubuntu-latest, but I'm still getting familiar with our CI, so could be wrong.

File: napari/benchmarks/benchmark_text_manager.py
Comment (andy-sweet) [inline napari/benchmarks/benchmark_text_manager.py:61]: If possible, links out to the official ASV docs (e.g. regarding to these parameters) or anything else would be useful here (permalink to ASV code, link to SO).

I guess something similar might be needed for any benchmark that mutates state in an additive/destructive way (e.g. adding or removing elements).

**Review by andy-sweet — COMMENTED** [review]
Overall, this looks great and I'm excited to merge it so that we can see what a run looks like!

I'm still learning about out CI, so you should definitely get someone else to look at this before merging, but hopefully some of my minor comments are helpful.

File: .github/workflows/test_comprehensive.yml
Comment (jaimergp) [reply .github/workflows/test_comprehensive.yml]: The Qt benchmarks do run on macos. We can use ubuntu for the non Qt parts. I added some comments to clarify this!

**Review by jni — APPROVED** [review]
This is very good stuff, @jaimergp! I agree re merging optimistically. Will merge tomorrow if no one objects. :heart: :zap:

**Review by tlambert03 — APPROVED** [review]

**Comment by jaimergp** [comment]
I think this is finally ready :)

**Comment by tlambert03** [comment]
thanks @jaimergp!  This is awesome.  Can you give a quick "how-to" summary here that we can refer to?  I know we can dig through the ci files to get this info, but would be nice to have: 

- how does this gets run (i gather it runs when one adds the `run-benchmarks` label to a PR)
- what is this actually benchmarking: is it just the PR vs main?
- is there a performance threshold that causes it to fail?
- is there any long-term regression reporting included here?  or will that be a later step?

**Comment by jaimergp** [comment]
Hey @tlambert03 thanks! I think it'd be better to provide that information as part of the documentation once we are finished with the full implementation. Right now I have only tested the PR runs (which require a label trigger, correct!), but we also have support for cron jobs and workflow_dispatch. Additionally, this includes a delegated PR to post a comment once the label CI ran, but we need that in `main` first (otherwise it won't work; permissions stuff).

So the idea is to merge this and then fix the rough edges we will for sure find. Once that's working, I'll put one more PR with the docs. What do you think?

**Pull request merged**
